### DISCUSSION TITLE
Converting to TypeScript

### DISCUSSION THREAD

**Discussion by apisaint** [Kong/insomnia#2989] [Show and Tell]
I've started a converting this repository from Flow to Typescript, but there is quite a lot of things that pose a challenge and honestly, I'm not sure whether or not it's even worth it at this point based on the initial start. 😅

Right now I have at least _[somewhat converted](https://github.com/apisaint/insomnia/tree/feat/typescript)_ the `insomnia-app` package to TypeScript, however, there is quite a bit of work still left:

- [ ] Convert files that error'd out by hand
- [ ] Fix issues from conversion (there are a lot)
- [ ] Replace converted Object types to `Record<string, any>`
- [ ] Move from `scripts` and `app` structure to `src/*` structure to create consistency
- [ ] Update build and start scripts to support typescript
- [ ] Update webpack for typescript
- [ ] Update babel for typescript
- [ ] Add `@types` for libraries

The above must be done for _each_ package.

Resources that I have been leveraging for guidance around monorepos and typescript:

- https://github.com/wwwebman/lerna-monorepo-typescript-example
- https://github.com/NiGhTTraX/ts-monorepo
- https://webman.pro/blog/how-to-setup-typescript-path-aliases-in-lerna-monorepo/
- https://skovhus.github.io/blog/flow-to-typescript-migration/

There is also the resource for electron:

- https://medium.com/@davembush/typescript-and-electron-the-right-way-141c2e15e4e1

**Comment by develohpanda** [comment]
Awesome notes, especially with the electron resources! :) I did a brief [proof-of-concept](https://github.com/Kong/insomnia/tree/TS) during some downtime as well, primarily using https://skovhus.github.io/blog/flow-to-typescript-migration/ and https://blog.logrocket.com/setting-up-a-monorepo-with-lerna-for-a-typescript-project-b6a81fe8e4f8/.

Instead of the `insomnia-app` package, I tried it with `openapi-2-kong`, because it has fewer dependencies and is well typed in comparison, just to test the approach. I did achieve success with that as well, so it's good validation that you've made some progress too! In addition to some of your bullet points we can also consolidate ES Lint, Jest and `tsconfig` files, which ultimately helped me with fixing errors. I'll add my notes when I find time to do so. 👍🏽 

You're right though, it's a pretty big undertaking, somewhat manual (since there are some repetitive changes for each package & plugin), but with the ts compiler and ES Lint it is certainly achievable.

We've also talked about this internally for a while now, and one of the items on the roadmap was to introduce some degree of E2E smoke tests before any major refactors, which are now in place (`packages/insomnia-smoke-test`) and more being added. 🎉

**Comment by wdawson** [comment]
# We've moved to TypeScript! 🎉 

As of 2021.4.0, the repository is (mostly) on TypeScript. Here was the plan of attack: #3654 

There's still a bit more to do, so don't hesitate to keep cleaning it up as you work on other issues. Great work by the team so far and we're excited to push this for our community ❤️
